### PR TITLE
[Kernel] Add B12X causal paged attention backend

### PR DESCRIPTION
**Pull request by lukealonso**
## Purpose

Builds on the optional B12X dependency and shared lazy-import integration merged in #52016.

This PR adds an explicitly selected
[B12X](https://github.com/local-inference-lab/b12x) causal paged-attention
backend for NVIDIA SM120 and SM121 GPUs using vLLM's existing attention backend
interface. It does not modify generic model-runner behavior or introduce a new
attention abstraction.

Supported paths include:

- Causal paged MHA, MQA, and GQA with BF16 model dtype.
- BF16 and FP8 E4M3 KV caches.
- Prefill, decode, and mixed batches.
- Speculative verification, sliding-window attention, and attention sinks.
- CUDA graph capture and replay.

MLA, non-causal attention, ALiBi, logits soft capping, and context parallelism
are unsupported.

The documentation covers explicit `--attention-backend B12X_ATTN` selection,
supported cache formats and head sizes, and unsupported attention modes.

This is the causal-attention component split from #51696, which it supersedes
rather than duplicates. Related open work targets different backend boundaries
and does not provide this standalone optional-package attention integration.

AI assistance from OpenAI Codex was used while developing this PR. I reviewed
every changed line and am responsible for understanding and defending the
integration end-to-end.

## Test Plan

Run the focused backend and reference-correctness coverage:

```bash
CUDA_VISIBLE_DEVICES=<idle-gpu> .venv/bin/python -m pytest \
  tests/v1/attention/test_b12x_attn.py \
  tests/v1/attention/test_attention_backends.py \
  -k b12x -q
```

Run pre-commit over every file changed by this PR layer:

```bash
mapfile -t changed_files < <(git diff --name-only upstream/main...HEAD)
.venv/bin/pre-commit run --files "${changed_files[@]}"
git diff --check upstream/main...HEAD
```

Benchmark `Qwen/Qwen3-8B` on an RTX PRO 6000 Blackwell Max-Q GPU with BF16
model weights, TP1, an 8192-token input, and a 512-token output. Hold linear
execution constant and compare B12X with FlashInfer using both BF16 and FP8 E4M3
KV caches. Use three warmup and ten measured requests per result.

## Test Result

Backend contract tests against the published `b12x==1.2.4` wheel:

```text
14 passed
```

GPU correctness and CUDA graph coverage:

```text
29 passed
```

All changed-file pre-commit hooks passed, and `git diff --check` reported no
errors.

Single-request end-to-end decode throughput; higher is better:

| Path | Model/configuration | TP | Comparison backend | Comparison tok/s | B12X tok/s | Change |
| --- | --- | ---: | --- | ---: | ---: | ---: |
| BF16-KV attention | Qwen3-8B, 8K context | 1 | FlashInfer | 88.5677 | 89.6471 | +1.22% |
| FP8-KV attention | Qwen3-8B, 8K context | 1 | FlashInfer | 91.0632 | 92.3668 | +1.43% |

The attention measurements exclude prefill and use 5,110 post-first-token
outputs across ten requests per result. All 40 measured requests completed
successfully.

---

<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR.
- [x] The test plan, including commands.
- [x] The test and performance results.
- [x] The necessary documentation update.

</details>

### CODE REVIEW

File: vllm/v1/attention/backends/registry.py [hunk 1]
```diff
@@ -94,6 +94,7 @@ class AttentionBackendEnum(Enum, metaclass=_AttentionBackendEnumMeta):
     ROCM_FLASHMLA_SPARSE_DSV4 = (
         "vllm.models.deepseek_v4.amd.rocm.DeepseekV4ROCMAiterMLASparseBackend"
     )
+    B12X_ATTN = "vllm.v1.attention.backends.b12x_attn.B12XPagedAttentionBackend"
```
Comment (lukealonso) [inline vllm/v1/attention/backends/registry.py]: This should probably just be "B12X", not B12X_ATTN

File: tests/v1/attention/test_attention_backends.py [hunk 2]
```diff
@@ -311,17 +316,19 @@ def mock_get_per_layer_parameters(vllm_config, layer_names, impl_cls):
...
-        scale=scale,
-        num_kv_heads=num_kv_heads,
-        alibi_slopes=None,
-        sliding_window=sliding_window,
-        attn_type=attn_type,
-        kv_cache_dtype=kv_cache_dtype,
-        **({"sinks": sinks} if sinks is not None else {}),
-    )
+    extra_impl_kwargs = {"sinks": sinks} if sinks is not None else {}
+    with set_current_vllm_config(vllm_config):
```
Comment (lukealonso) [inline tests/v1/attention/test_attention_backends.py]: Going to refactor this so touching these other tests becomes unnecessary

File: docs/features/quantization/b12x.md [hunk 3]
```diff
@@ -1,24 +1,34 @@
-# B12X Linear Backend
+# b12x Linear and Attention Backends
```
Comment (lukealonso) [inline docs/features/quantization/b12x.md]: Does this really belong in a quantization document? The linears and moe I can understand, but the attention backend really has nothing to do with quantization.

File: tests/v1/attention/test_attention_backends.py [hunk 4]
```diff
@@ -337,9 +342,28 @@ def mock_get_per_layer_parameters(vllm_config, layer_names, impl_cls):
...
             mock_layer, key, value, kv_cache, attn_metadata.slot_mapping
         )
-    output = impl.forward(
-        mock_layer, query, key, value, kv_cache, attn_metadata, output=output
-    )
+    if use_cuda_graph:
+        impl.forward(
+            mock_layer, query, key, value, kv_cache, attn_metadata, output=output
+        )
+        torch.accelerator.synchronize()
```
Comment (lukealonso) [inline tests/v1/attention/test_attention_backends.py]: What's with the device synchronize here? Leave a comment at least. Also, is this how other tests handle mocking cuda graphs?

File: tests/v1/attention/test_attention_backends.py [hunk 5]
```diff
@@ -589,23 +646,30 @@ def _test_backend_correctness(
                     kv_cache.permute(*stride_order).contiguous().permute(*inv_order)
                 )
 
+        config_context = (
```
Comment (lukealonso) [inline tests/v1/attention/test_attention_backends.py]: Why is this backend special such that it needs this config context support?

File: tests/v1/attention/test_attention_backends.py [hunk 6]
```diff
@@ -637,6 +701,150 @@ def error_msg(msg: str, backend_name: str):
...
+):
+    """b12x preserves causal SWA and attention-sink semantics."""
+    _require_b12x_paged_attention()
+
+    sliding_window = 16
+    mask = partial(_b12x_causal_sliding_window_mask, sliding_window=sliding_window)
+
+    _test_backend_correctness(
+        BATCH_SPECS[batch_spec_name],
+        "Qwen/Qwen3-0.6B",
```
Comment (lukealonso) [inline tests/v1/attention/test_attention_backends.py]: This isn't a SWA model, is this sensible? How do other SWA backends perform this test?

File: tests/v1/attention/test_attention_backends.py [hunk 7]
```diff
@@ -637,6 +701,150 @@ def error_msg(msg: str, backend_name: str):
...
+def test_b12x_decode_cuda_graph_replay_with_sliding_window_and_sinks(
+    default_vllm_config,
+    workspace_init,
+    kv_cache_dtype: str,
+):
+    """b12x causal metadata and sinks remain capture-safe during replay."""
+    _require_b12x_paged_attention()
+
+    sliding_window = 16
+    mask = partial(_b12x_causal_sliding_window_mask, sliding_window=sliding_window)
```
Comment (lukealonso) [inline tests/v1/attention/test_attention_backends.py]: Same here.

File: tests/v1/attention/test_attention_backends.py [hunk 8]
```diff
@@ -637,6 +701,150 @@ def error_msg(msg: str, backend_name: str):
...
+
+
+def test_b12x_speculative_verifier_cuda_graph_replay(
+    default_vllm_config,
+    workspace_init,
+):
+    """b12x replays uniform speculative verification through its graph plan."""
+    _require_b12x_paged_attention()
+
+    _test_backend_correctness(
```
Comment (lukealonso) [inline tests/v1/attention/test_attention_backends.py]: Is this a normal thing to test? (speculative verification specifically - what's special about this vs any other forward pass?)

**Review by LucasWilkinson — APPROVED** [review]
LGTM
